src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java

```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        }
+        else throw new SlimFixtureException("There are only "+ messageAttachments.size() + " attachments");
+    }
+
+    public String saveAttachmentByFilename(String filename)
+    {
+        for(ImapAttachment ia: messageAttachments) {
+            if(StringUtils.equals(ia.getFileName(), filename)) return impSaveAttachment(ia);
+        }
+        throw new SlimFixtureException("There is no file with" + filename + " attachments");
+    }
+    private String impSaveAttachment(ImapAttachment ia) {
+        try {
+            return createFile(attachmentBase, ia.getFileName(), ia.getBytes());
+        } catch (IOException ex) {
+            throw new RuntimeException("Unable to readstream", ex);
+        }
+    }
+
     public String body() {
         String result = bodyText();
         return getEnvironment().getHtml(result);
     }
 
     public String bodyText() {
-        return applyToLastMessage(this::getBody);
+        return messageHtml;
+    }
+
+    public String bodyPlain() {
+        return messagePlain;
+    }
+
+    private void handleParts(Part part) {
+        try {
+            if (part == null) return;
+            String disposition = part.getDisposition();
+            if (Part.ATTACHMENT.equalsIgnoreCase(disposition)) {
+                messageAttachments.add(new ImapAttachment(part));
+            } else {
+                Object content = part.getContent();
+                if (content instanceof MimeMultipart) {
+                    Multipart multipart = (Multipart) content;
+                    for (int i = 0; i < multipart.getCount(); i++) {
+                        handleParts(multipart.getBodyPart(i));
+                    }
+                } else if (content instanceof String) {
+                    if (part.isMimeType("text/html")) messageHtml = (String)content;
+                    else messagePlain = (String)content;
+                } else throw new RuntimeException("Unknown content type " + content.getClass());
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException("Unable to get body of message", ex);
+        } catch (MessagingException ex) {
+            throw new RuntimeException("Unable to get body of message", ex);
+        }
     }
 
     public boolean retrieveLastMessage() {
         lastMessage = null;
+        Message[] messages = retrieveMessages();
+        boolean result = (messages != null && messages.length > 0);
+        setLastMessage(result ? messages[messages.length - 1]: null  );
+        return result;
+    }
+
+    protected Message[] retrieveMessages() {
         SearchTerm searchTerm = getSearchTerm();
         try {
-            Message[] messages = openFolder().search(searchTerm);
-            boolean result = messages.length > 0;
-            if (result) {
-                lastMessage = messages[messages.length - 1];
-            }
-            return result;
+            Folder f = openFolder();
```

Comment:
Why only close `f` if there are no messages? Shouldn't that always be done? `Folder` is `AutoClosable` so could be placed inside a `try`:
`try (Folder f = openFolder()) {`

Reply:
Ill have to check if this doesn't create issues with handleParts, impSaveAttachment & moveMessage